[ZeroGPU] Dynamic duration + compilation (#1896)
huggingface/hub-docs: docs/hub/spaces-zerogpu.md

### Supported Versions

- **Gradio**: 4+

> - **PyTorch**: Almost all versions from **2.1.0** to **latest** are supported  

Reviewer: Are these all the versions except 2.3.x? If so, we could say it like that.

This sets the maximum function runtime to 120 seconds. Specifying shorter durations for quicker functions will improve queue priority for Space visitors.

### Dynamic duration

`@spaces.GPU` also supports dynamic durations.

Instead of directly passing a duration, simply pass a callable that takes the same inputs as your decorated function and returns a duration:

```python
def get_duration(prompt, steps):
    step_duration = 3.75
    return steps * step_duration

@spaces.GPU(duration=get_duration)
def generate(prompt, steps):
   return pipe(prompt, num_inference_steps=steps).images
```

## Compilation

> ZeroGPU does not support `torch.compile` but you can use PyTorch **ahead-of-time** compilation.

Reviewer 2: maybe mention the pytorch version for which onwards this is supported as well

This sets the maximum function runtime to 120 seconds. Specifying shorter durations for quicker functions will improve queue priority for Space visitors.

### Dynamic duration

`@spaces.GPU` also supports dynamic durations.

> Instead of directly passing a duration, simply pass a callable that takes the same inputs as your decorated function and returns a duration:

Reviewer: ```suggestion
Instead of directly passing a duration, simply pass a callable that takes the same inputs as your decorated function and returns a duration value:
```

```python
def get_duration(prompt, steps):
    step_duration = 3.75
    return steps * step_duration

@spaces.GPU(duration=get_duration)
def generate(prompt, steps):
   return pipe(prompt, num_inference_steps=steps).images
```

## Compilation

> ZeroGPU does not support `torch.compile` but you can use PyTorch **ahead-of-time** compilation (requires torch `2.8+`)

Reviewer: ```suggestion
ZeroGPU does not support `torch.compile`, but you can use PyTorch **ahead-of-time** compilation (requires torch `2.8+`)
```

> Checkout this [blogpost](https://huggingface.co/blog/zerogpu-aoti) for a complete compilation guide on ZeroGPU

Reviewer: ```suggestion
Check out this [blogpost](https://huggingface.co/blog/zerogpu-aoti) for a complete guide on ahead-of-time compilation on ZeroGPU.
```
